JIT: Fold more casts

EgorBo opened the pull request.

EgorBo committed:
Move even more casts to the late phase

EgorBo committed:
fix naot

EgorBo commented:
@MihuBot

EgorBo committed:
Clean up

EgorBo committed:
Clean up

EgorBo committed:
clean up

EgorBo commented:
@MihuBot

EgorBo commented:
PTAL @jakobbotsch @AndyAyersMS cc @dotnet/jit-contrib 
this folds more casts (mostly, in tests);

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=568961&view=ms.vss-build-web.run-extensions-tab)

jakobbotsch commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -2704,6 +2701,42 @@ GenTree* Compiler::optVNBasedFoldExpr_Call(BasicBlock* block, GenTree* parent, G
+                if (castResult == TypeCompareState::Must)
+                {
+                    GenTree* result = gtWrapWithSideEffects(castObjArg, call, GTF_ALL_EFFECT, true);
+                    fgSetTreeSeq(result);
```
Is this necessary?

EgorBo replied:
Removed (and slightly refactored)

EgorBo committed:
Clean up (move logic to vn)

EgorBo committed:
Formatting

EgorBo committed:
Introduce

EgorBo commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -18500,6 +18500,7 @@ bool Compiler::gtStoreDefinesField(
 //    tree -- tree to find handle for
 //    pIsExact   [out] -- whether handle is exact type
 //    pIsNonNull [out] -- whether tree value is known not to be null
+//    vnBased          -- can we rely on value numbering to find the handle?
```
Introduced `fgVnValid` for it, going to think about using it in various asserts.

SingleAccretion replied:
`vnStore != nullptr` could be used as a check, it would avoid an additional boolean.

EgorBo replied:
> `vnStore != nullptr` could be used as a check, it would avoid an additional boolean.

@SingleAccretion does vnStore becomes null once VN are stale?

EgorBo replied:
Looks like it's not. It's only to null set in ResetOptAnnotations (which is JitOptRepeat specific)

SingleAccretion replied:
Yes, but you could null it out just like you're setting `fgVnValid` to `false`. It would "self-validate" that nothing uses it that way as well.

EgorBo replied:
Done (with a small fix in gtExtractSideEffList which was called in late phases)

EgorBo committed:
Address feedback

EgorBo committed:
Fix gtExtractSideEffList

EgorBo commented:
@AndyAyersMS @jakobbotsch I've addressed your concerns, PTAL again, thanks!

I slightly refactored it by moving most of the logic to `VNForCast` so if `VNForCast` assigns the same VN as the object being casted has - it means the cast is no-op (since there is no VNF_InvalidCastExc attached). It became possible after @jakobbotsch merged https://github.com/dotnet/runtime/pull/98417, previously, it caused weird CSE size regressions.

EgorBo commented:
@MihuBot

EgorBo commented:
Ping @AndyAyersMS @jakobbotsch 🙂

AndyAyersMS reviewed (commented):
The flow here is a little a bit convoluted, though I think we can live with it.

I wonder how hard it would be to figure this out "purely" on the VN side of things, and if not being able to do this means there is some missing modelling in VN.

EgorBo commented:
> I wonder how hard it would be to figure this out "purely" on the VN side of things, and if not being able to do this means there is some missing modelling in VN.

Yes, it can be cleaned up if we make the purely-VN `GetObjectType` on par with tree-based `gtGetClassHandle`, but since the latter is quite big alreayd I assume it will be a bit of work to do.

EgorBo commented on src/coreclr/jit/valuenum.cpp:
```diff
@@ -2464,6 +2464,87 @@ ValueNum ValueNumStore::VNForFunc(var_types typ, VNFunc func, ValueNum arg0VN)
+        return VNWithExc(objVN, vnExcSet);
+    }
+
+    // IsInstanceOf(cls, obj) -> either obj or null - we don't know
```
Good point, should simplify code and maybe improve the TP a bit

EgorBo replied:
Ah, actually, the caller of this function already relies on its result to add into the VN cache, so if I call VNForFuncNoFolding I'll end up hitting an assert on an attempt to add a VN to the cache where it's already presented 🙂

EgorBo commented on src/coreclr/jit/valuenum.cpp:
```diff
@@ -2464,6 +2464,87 @@ ValueNum ValueNumStore::VNForFunc(var_types typ, VNFunc func, ValueNum arg0VN)
+    //
+    // Fold "CAST(IsInstanceOf(obj, cls), cls)" to "IsInstanceOf(obj, cls)"
+    // where CAST is either ISINST or CASTCLASS. Most cases like this are expected to be handled
+    // by gtGetClassHandle above, except the casts involving runtime lookups.
```
Yes, it is - I'll just remove it

EgorBo committed:
Address feedback

EgorBo committed:
revert the change
